### PR TITLE
Add Pytorch Lightning spark estimator

### PR DESCRIPTION
**Pull request by irasit**
TorchEstimator(model) can be either a traditional PyTorch module or a PyTorch Lightning LightningModule.

When PyTorch Lightning is used, the optimizer and loss params must be omitted.

## Checklist before submitting

- [x] Did you read the [contributor guide](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md)?
- [x] Did you update the docs?
- [x] Did you write any tests to validate this change?  
- [x] Did you update the [CHANGELOG](https://github.com/horovod/horovod/blob/master/CHANGELOG.md), if this change affects users?

## Description

Fixes # (issue).

## Review process to land 

1. All tests and other checks must succeed.
2. At least one member of the [technical steering committee](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md) must review and approve.
3. If any member of the technical steering committee requests changes, they must be addressed.

### CODE REVIEW

**Comment by EnricoMi** [comment]
You are running a lot of test here. To reduce turnaround time on your side and costs on our side, please consider reducing the test space to test images that are most affected by your changes, e.g. different Spark versions and GPU.

For this you can remove some lines from here until your code is complete:
https://github.com/horovod/horovod/blob/6f0bb9fae826167559501701d4a5a0380284b5f0/.buildkite/gen-pipeline.sh#L20-L59

And the run

    BUILDKITE_PIPELINE_SLUG=SLUG BUILDKITE_BRANCH=BRANCH .buildkite/gen-pipeline.sh > test/single/data/expected_buildkite_pipeline.yaml

Commit this and there will be fewer tests.

Remember to revert this commit later before merging into master, which will finally run all tests again.

File: .buildkite/gen-pipeline.sh [hunk 1]
```diff
@@ -361,6 +361,10 @@ run_spark_integration() {
     run_test "${test}" "${queue}" \
       ":spark: Spark Torch MNIST (${test})" \
       "bash -c \"OMP_NUM_THREADS=1 python /horovod/examples/spark/pytorch/pytorch_spark_mnist.py --num-proc 2 --work-dir /work --data-dir /data --epochs 3\""
+
+    run_test "${test}" "${queue}" \
+      ":spark: Spark Torch MNIST (${test})" \
```
Comment (EnricoMi) [inline .buildkite/gen-pipeline.sh:366]: the test title should be unique, please add 'Lightning' it.

File: .travis.yml [hunk 2]
```diff
@@ -44,6 +47,7 @@ install:
   - python -m pip install -U pip
   - pip install tensorflow==${TENSORFLOW} keras==${KERAS}
   - pip install torch==${PYTORCH} torchvision==${TORCHVISION}
```
Comment (EnricoMi) [inline .travis.yml:49]: you can append `pytorch_lighning` to the line above, as you do in `Dockerfile.cpu`

File: .travis.yml [hunk 3]
```diff
@@ -18,6 +18,7 @@ jobs:
         - KERAS=2.2.4
         - PYTORCH=1.2.0
         - TORCHVISION=0.4.0
+        - PYTORCH_LIGHTNING=0.7.5
```
Comment (EnricoMi) [inline .travis.yml:21]: Can you move `PYTORCH_LIGHTNING' one line up, so it aligns with `PYTORCH`?

File: docker-compose.test.yml [hunk 4]
```diff
@@ -13,6 +13,7 @@ services:
         KERAS_PACKAGE: keras==2.4.3
         PYTORCH_PACKAGE: torch==1.8.1+cpu
         TORCHVISION_PACKAGE: torchvision==0.9.1+cpu
+        PYTORCH_LIGHTNING: pytorch_lightning==1.2.6
```
Comment (EnricoMi) [inline docker-compose.test.yml:16]: Can you move `PYTORCH_LIGHTNING` below `PYTORCH` here as well?

File: docker-compose.test.yml [hunk 5]
```diff
@@ -135,6 +137,7 @@ services:
         KERAS_PACKAGE: keras==2.3.1
         PYTORCH_PACKAGE: torch==1.6.0+cu101
         TORCHVISION_PACKAGE: torchvision==0.7.0+cu101
+        PYTORCH_LIGHTNING: pytorch_lightning==1.2.6
```
Comment (EnricoMi) [inline docker-compose.test.yml:140]: Are there cuda versions for pytorch_lightning, or is this pure python on top of the respective pytorch package (which has the cuda bit)?

File: horovod/spark/common/params.py [hunk 6]
```diff
@@ -53,8 +53,7 @@ class EstimatorParams(Params):
     label_cols = Param(Params._dummy(), 'label_cols', 'label column names',
                        typeConverter=TypeConverters.toListString)
     validation = Param(Params._dummy(), 'validation',
-                       'one of: float validation split [0, 1), or string validation column name',
-                       typeConverter=TypeConverters.toString)
```
Comment (EnricoMi) [inline horovod/spark/common/params.py:57]: why is the string converter removed?

File: horovod/spark/common/store.py [hunk 7]
```diff
@@ -83,7 +83,11 @@ def get_run_path(self, run_id):
         raise NotImplementedError()
 
     def get_checkpoint_path(self, run_id):
-        """Returns the path to the checkpoint file for the given run."""
+        """Returns the path to the checkpoint file(s) for the given run."""
+        raise NotImplementedError()
+
+    def get_checkpoints(self, run_id, suffix='.ckpt'):
+        """Returns a list of paths for all chceckpoints saved this run."""
```
Comment (EnricoMi) [inline horovod/spark/common/store.py:90]: typo `chceckpoints`

File: horovod/spark/common/store.py [hunk 8]
```diff
@@ -231,12 +235,17 @@ def get_checkpoint_path(self, run_id):
...
+        checkpoint_dir = self.get_localized_path(self.get_checkpoint_path(run_id))
+        filenames = self.get_filesystem().ls(checkpoint_dir)
+        return sorted([name for name in filenames if name.endswith(suffix)])
+
     def get_logs_path(self, run_id):
         return os.path.join(self.get_run_path(run_id), self.get_logs_subdir()) \
             if self._save_runs else None
 
     def get_checkpoint_filename(self):
-        return 'checkpoint.h5'
```
Comment (EnricoMi) [inline horovod/spark/common/store.py:239]: @tgaddair is this a breaking change?

File: test/single/test_ray.py [hunk 9]
```diff
@@ -28,7 +28,7 @@ def ray_start_2_cpus():
 
 @pytest.fixture
 def ray_start_4_cpus():
-    address_info = ray.init(num_cpus=4)
+    address_info = ray.init(num_cpus=4, _redis_max_memory=1024*1024*1024)
```
Comment (EnricoMi) [inline test/single/test_ray.py:31]: does this fix some specific transient ray test errors?

**Review by EnricoMi — CHANGES_REQUESTED** [review]

File: test/single/test_ray.py (hunk 9 above)
Comment (irasit) [reply test/single/test_ray.py:31]: It fix the case in local. But the root cause of buildkite failure is the lightning version missmatch.

File: horovod/spark/common/store.py (hunk 8 above)
Comment (irasit) [reply horovod/spark/common/store.py:239]: It was, I fixed it later.

File: horovod/spark/common/params.py (hunk 6 above)
Comment (irasit) [reply horovod/spark/common/params.py:57]: validation: Optional validation column name (string) where every row in the column is either 1/True or 0/False,
                    or validation split (float) giving percent of data to be randomly selected for validation.

Comment (irasit) [reply horovod/spark/common/params.py:57]: It can be a string for col name or float for split.

File: docker-compose.test.yml (hunk 5 above)
Comment (irasit) [reply docker-compose.test.yml:140]: This is pure python on top, it will relay on horovod backend and torch for GPU training.

File: .travis.yml (hunk 3 above)
Comment (irasit) [reply .travis.yml:21]: sure

File: docker-compose.test.yml (hunk 4 above)
Comment (irasit) [reply docker-compose.test.yml:16]: sure

File: .travis.yml (hunk 2 above)
Comment (irasit) [reply .travis.yml:49]: sure

File: .buildkite/gen-pipeline.sh (hunk 1 above)
Comment (irasit) [reply .buildkite/gen-pipeline.sh:366]: name changed.

File: test/single/test_ray.py (hunk 9 above)
Comment (tgaddair) [reply test/single/test_ray.py:31]: I've also seen some random flakiness with ray in tests that requires changing these params. I think this is fine if it unblocks for now.

**Review by tgaddair — APPROVED** [review]
This is awesome!  Nice work @irasit and @chongxiaoc!

**Comment by tgaddair** [comment]
@EnricoMi does everything look good to you?

File: .buildkite/gen-pipeline.sh [hunk 10]
```diff
@@ -74,7 +74,7 @@ build_test() {
   echo "      push-retries: 5"
   echo "  - ecr#v1.2.0:"
   echo "      login: true"
-  echo "  timeout_in_minutes: 30"
+  echo "  timeout_in_minutes: 60"
```
Comment (EnricoMi) [inline .buildkite/gen-pipeline.sh:77]: Do build times really go up this much with your changes?

File: examples/pytorch/pytorch_lightning_mnist.py [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+    if hvd.rank() == 0:
+        print('\nTest set: Average loss: {:.4f}, Accuracy: {:.2f}%\n'.format(
+            test_loss, 100. * test_accuracy))
+
+if __name__ == '__main__':
+    args = parser.parse_args()
+    args.cuda = not args.no_cuda and torch.cuda.is_available()
+    hvd.init()
+
+    #kwargs = {'num_workers': 1, 'pin_memory': True} if args.cuda else{}
```
Comment (EnricoMi) [inline examples/pytorch/pytorch_lightning_mnist.py:127]: this commented code should be removed

File: horovod/spark/common/params.py (hunk 6 above)
Comment (EnricoMi) [reply horovod/spark/common/params.py:57]: and with that converter it always turned it into strings?

File: horovod/spark/common/store.py (hunk 8 above)
Comment (EnricoMi) [reply horovod/spark/common/store.py:239]: @tgaddair?

File: Dockerfile.cpu [hunk 12]
```diff
@@ -3,6 +3,7 @@ FROM ubuntu:18.04
 ENV TENSORFLOW_VERSION=2.3.0
 ENV PYTORCH_VERSION=1.6.0
 ENV TORCHVISION_VERSION=0.7.0
+ENV PYTORCH_LIGHTNING_VERSION=1.2.6
```
Comment (EnricoMi) [inline Dockerfile.cpu:6]: can you move this one line up?

File: Dockerfile.gpu [hunk 13]
```diff
@@ -4,6 +4,7 @@ FROM nvidia/cuda:10.1-devel-ubuntu18.04
 ENV TENSORFLOW_VERSION=2.3.0
 ENV PYTORCH_VERSION=1.6.0
 ENV TORCHVISION_VERSION=0.7.0
+ENV PYTORCH_LIGHTNING_VERSION=1.2.6
```
Comment (EnricoMi) [inline Dockerfile.gpu:7]: can you move this one line up?

File: Dockerfile.test.cpu [hunk 14]
```diff
@@ -12,6 +12,7 @@ ARG TENSORFLOW_PACKAGE=tensorflow-cpu==1.15.0
 ARG KERAS_PACKAGE=keras==2.2.4
 ARG PYTORCH_PACKAGE=torch==1.2.0+cpu
 ARG TORCHVISION_PACKAGE=torchvision==0.4.0+cpu
+ARG PYTORCH_LIGHTNING_PACKAGE=pytorch_lightning==0.7.6
```
Comment (EnricoMi) [inline Dockerfile.test.cpu:15]: can you move this one line up?

File: Dockerfile.test.gpu [hunk 15]
```diff
@@ -13,6 +13,7 @@ ARG TENSORFLOW_PACKAGE=tensorflow-gpu==1.15.0
 ARG KERAS_PACKAGE=keras==2.2.4
 ARG PYTORCH_PACKAGE=torch==1.2.0
 ARG TORCHVISION_PACKAGE=torchvision==0.4.0
+ARG PYTORCH_LIGHTNING_PACKAGE=pytorch_lightning==0.7.6
```
Comment (EnricoMi) [inline Dockerfile.test.gpu:16]: can you move this one line up?

**Review by EnricoMi — COMMENTED** [review]
some minor open questions and changes

File: .buildkite/gen-pipeline.sh (hunk 10 above)
Comment (irasit) [reply .buildkite/gen-pipeline.sh:77]: Saw it used more time, but could be one time spike. Will roll it back and try

**Comment by irasit** [comment]
Fixed all the comment. Seems pytorch-lightning 1.2.8 dose not contain the hvd.join(local_rank) fix. Need to wait for the next version.

**Review by EnricoMi — APPROVED** [review]
Excellent, approved, When the release has happened, please upgrade to get the tests green.

Can you turn this into a draft for the time being, so we do not accidentally merge it into master and break tests?

**Comment by EnricoMi** [comment]
I have turned the PR into a draft until the release.
